app/src/main/java/com/sopt/now/presentation/Login/LoginViewModel.kt

```diff
@@ -0,0 +1,58 @@
+package com.sopt.now.presentation.Login
+
+
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.sopt.now.domain.AuthRepository
+import com.sopt.now.domain.model.AuthData
+import kotlinx.coroutines.launch
+import retrofit2.HttpException
+
+class LoginViewModel(
+    private val authRepository: AuthRepository
+) : ViewModel() {
+    val loginResult: MutableLiveData<Boolean> = MutableLiveData()
+    val errorMessage: MutableLiveData<String> = MutableLiveData()
+    val userId: MutableLiveData<String> = MutableLiveData()
+    fun logIn(authData: AuthData) {
+        viewModelScope.launch{
+            authRepository.logIn(authData)
+            .onSuccess { response ->
+                val userId =  response.headers()["Location"]
+                userId?.let {
+                    this@LoginViewModel.userId.value =it
+                    loginResult.value = true
+                }
+            }
+            .onFailure{ t ->
+                if(t is HttpException){
+                    val error = t.message()
+                    errorMessage.postValue("$error")
+                }
+                errorMessage.postValue("서버 에러 발생")}
+
```

Comment:
이 경우 서버에러가 아닌, timeout, json 파싱, 네트워크 연결 끊김 등등의 오류가 throw되는데 "서버 에러 발생" 으로 네이밍 짓게 되면 사용자에게 오히려 혼란을 줄 수 있겠죠?

무엇보다 지금 구조에서는 if문이 끝나고  errorMessage.postValue("서버 에러 발생") 가 반드시 실행 되기 때문에 postValue가 두번 이루어지는 상황입니다! 이렇게 되면 HttpException시 서버에서 내려준 오류 메세지(잘못된 비밀 번호입니다.. 등등) 이 안보이고 무조건 서버 에러 발생이 보이겠죠?

```suggestion
            .onFailure{ t ->
                if(t is HttpException){
                    val error = t.message()
                    errorMessage.postValue("$error")
                } else {
                    errorMessage.postValue(t.message.toString())
                }
```
이런식으로 분기처리를 해주고 여러가지 오류 상황들을 toast로 띄우는건 어떨까요?

Reply:
추가로 이러한 HttpException은 외부 라이브러리 retrofit에 의존하게 됩니다!
ui영역은 이러한 서버 통신 라이브러리에 의존하는 것은 좋지 않습니다! 특히나 viewmodel은 더욱더요

이번 과제를 통해 data영역에서 repository를 통해 서버 통신을 data layer로 잘 분리 하셨습니다.

그렇다면 아키텍쳐 관점에서 실제 AuthService가 불리는 data 영역의 RepoImpl에서 이러한 Error handling 작업을 하는 것이 더 낫겠죠?

AuthRepoImpl 에서 error handling을 하는 것이 좋겠죠? 근데 이거 6차 세미나 자료에 있었는데 왜 사라졌지..?
onFailure블록을 추가하고 여기서 Result.Failure(에러메세지)를 담아 보내면 됩니다!

이 부분 해보시고 안되면 물어보세요!